File: 백준/Gold/1918. 후위 표기식/후위 표기식.py
chars = list(input())
prior = {
    '(':1, 
    '+':2, 
    '-':2, 
    '*':3, 
    '/':3, 
    ')':4}
stack = []
result = []
operator = ['+', '-', '*', '/']

for char in chars:
    if char == '(':
        stack.append(char)
    elif char in operator:
        while stack and prior[stack[-1]] >= prior[char]:
            result.append(stack.pop())
        stack.append(char)
    elif char == ')':
        while True:
            temp = stack.pop()
            if temp == '(':
                break
            result.append(temp)
    else:
        result.append(char)

while stack:
    result.append(stack.pop())

print(''.join(result))

File: 백준/Gold/1918. 후위 표기식/후위 표기식.js
const fs = require("fs");
const input = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

const infix = input[0].trim().split("");
let answer = "";

const stack = [];
while (infix.length > 0) {
  const now = infix.shift();

  if (now === "(") {
    stack.push("(");
  } else if (now === ")") {
    while (stack.length > 0 && stack[stack.length - 1] !== "(") {
      answer += stack.pop();
    }
    stack.pop(); // ( 빼주기
  } else if (now === "+" || now === "-") {
    while (stack.length > 0 && stack[stack.length - 1] !== "(") {
      answer += stack.pop();
    }
    stack.push(now);
  } else if (now === "*" || now === "/") {
    while (
      stack.length > 0 &&
      (stack[stack.length - 1] === "*" || stack[stack.length - 1] === "/")
    ) {
      answer += stack.pop();
    }
    stack.push(now);
  } else {
    answer += now;
  }
}

while (stack.length > 0) {
  answer += stack.pop();
}
console.log(answer);
